source/extensions/http/header_validators/envoy_default/http2_header_validator.cc

```diff
@@ -0,0 +1,415 @@
+#include "source/extensions/http/header_validators/envoy_default/http2_header_validator.h"
+
+#include "source/extensions/http/header_validators/envoy_default/character_tables.h"
+
+#include "absl/container/node_hash_map.h"
+#include "absl/container/node_hash_set.h"
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Http {
+namespace HeaderValidators {
+namespace EnvoyDefault {
+
+using ::envoy::extensions::http::header_validators::envoy_default::v3::HeaderValidatorConfig;
+using ::Envoy::Http::HeaderString;
+using ::Envoy::Http::LowerCaseString;
+using ::Envoy::Http::Protocol;
+using HeaderValidatorFunction = ::Envoy::Http::HeaderValidator::HeaderEntryValidationResult (
+    Http2HeaderValidator::*)(const HeaderString&);
+
+struct Http2ResponseCodeDetailValues {
+  const std::string InvalidTE = "uhv.http2.invalid_te";
+  const std::string ConnectionHeaderSanitization = "uhv.http2.connection_header_rejected";
+};
+
+using Http2ResponseCodeDetail = ConstSingleton<Http2ResponseCodeDetailValues>;
+
+/*
+ * Header validation implementation for the Http/2 codec. This class follows guidance from
+ * several RFCS:
+ *
+ * RFC 3986 <https://datatracker.ietf.org/doc/html/rfc3986> URI Generic Syntax
+ * RFC 9110 <https://www.rfc-editor.org/rfc/rfc9110.html> HTTP Semantics
+ * RFC 9112 <https://www.rfc-editor.org/rfc/rfc9112.html> HTTP/1.1
+ * RFC 9113 <https://www.rfc-editor.org/rfc/rfc9113.html> HTTP/2
+ *
+ */
+Http2HeaderValidator::Http2HeaderValidator(const HeaderValidatorConfig& config, Protocol protocol,
+                                           StreamInfo::StreamInfo& stream_info)
+    : HeaderValidator(config, protocol, stream_info) {}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http2HeaderValidator::validateRequestHeaderEntry(const HeaderString& key,
+                                                 const HeaderString& value) {
+  static const absl::node_hash_map<absl::string_view, HeaderValidatorFunction> kHeaderValidatorMap{
+      {":method", &Http2HeaderValidator::validateMethodHeader},
+      {":authority", &Http2HeaderValidator::validateAuthorityHeader},
+      {":scheme", &Http2HeaderValidator::validateSchemeHeader},
+      {":path", &Http2HeaderValidator::validatePathHeaderCharacters},
+      {"te", &Http2HeaderValidator::validateTEHeader},
+      {"content-length", &Http2HeaderValidator::validateContentLengthHeader},
```

Comment:
Presently H/1 and H/2 codecs reject requests with multiple content-length headers or a multivalue content-length header. In the case of header validator, multiple content-length headers will be concatenated into one multivalue header and then fail parsing it as a number because of comma.

The standard does allow multiple content-length values that are the same though, leaving it up to the implementation on how to deal with this. Rejecting it is legal which is what we are doing now.

```
   a recipient of a Content-Length header field value
   consisting of the same decimal value repeated as a comma-separated
   list (e.g, "Content-Length: 42, 42") MAY either reject the message as
   invalid or replace that invalid field value with a single instance of
   the decimal value, since this likely indicates that a duplicate was
   generated or combined by an upstream message processor.
```

@ameily I think it is worth adding a comment to the `validateContentLengthHeader` that the standard allows `Content-Length: 42, 42` and that an option to accept this may be added in the future.